utils/generic.py

```diff
@@ -145,24 +148,31 @@ def split_multi_dynamic(t: Union[hl.MatrixTable, hl.Table], keep_star: bool = Fa
     return sm.result()
 
 
-def pc_project(mt: hl.MatrixTable, pc_loadings: hl.Table,
-               loading_location: str = "loadings", af_location: str = "pca_af") -> hl.MatrixTable:
+def pc_project(mt: hl.MatrixTable, loading_location: str = "loadings", af_location: str = "pca_af") -> hl.Table:
     """
-    Projects samples in `mt` on PCs computed in `pc_mt`
+    Projects samples in `mt` on pre-computed PCs
+
+    The input MatrixTable should have a column with the pca loadings and another with the allele frequencies
+    used for the PCA.
+
     :param MatrixTable mt: MT containing the samples to project
-    :param Table pc_loadings: MT containing the PC loadings for the variants
-    :param str loading_location: Location of expression for loadings in `pc_loadings`
-    :param str af_location: Location of expression for allele frequency in `pc_loadings`
-    :return: MT with scores calculated from loadings
+    :param str loading_location: Location of expression for loadings in `mt`
+    :param str af_location: Location of expression for allele frequency in `mt`
+    :return: Table with scores calculated from loadings in column `scores`
+    :rtype: Table
     """
-    n_variants = mt.count_rows()
-
-    mt = mt.annotate_rows(**pc_loadings[mt.locus, mt.alleles])
     mt = mt.filter_rows(hl.is_defined(mt[loading_location]) & hl.is_defined(mt[af_location]) &
                         (mt[af_location] > 0) & (mt[af_location] < 1))
 
+    mt = mt.persist()
+
+    n_variants = mt.count_rows()
```

Comment:
Are you sure this is the correct order? I can understand this after filtering for `hl.is_defined(mt[loading_location]) & hl.is_defined(mt[af_location])`, but if your new dataset is monomorphic at any of the sites, that site should probably be counted, no? This count should be the same as the number of variants used for PCA, and should not be specific to the projected dataset, I'd think? I could be wrong.

Reply:
I think you're right. so in fact, before should have been pc_loadings.count_rows() and not mt.count_rows(). But thinking about it, it makes sense indeed. So now I'm less thrilled about not passing the loadings table any more, I'll re-work that part.